### PR TITLE
Add stake_flags to stake state

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

Fix a stake redelegation issue.

Part of https://github.com/solana-labs/solana/pull/31876

#### Summary of Changes

Add StakeFlags to StakeState

NOTE: This change just adds StakeFlags to StakeState. The Stake behavior should not change. 
The actual fix will come in a following PR once this PR is merged.  

Fixes #

### CODE REVIEW

**Review by t-nelson — COMMENTED** [review]
lgtm. gonna give spl folks the final approval here though since you'll need to workout deployment strategy with them

**Comment by mvines** [comment]
We probably should bump the major version of the `solana-program` crate with this PR (once releng folks figure out how to do this!) since it's a breaking change

**Comment by t-nelson** [comment]
> I am not familiar with the version of `solana-program`. Do you mean to bump the following version?
> 
> `./stake-pool/program/Cargo.toml:version = "0.7.0"`

no. the `version` field in `sdk/program/Cargo.toml` (monorepo).  there's an initiate to move both this crate and `solana-sdk` out of the monorepo, so they can be more easily independently versioned

**Comment by t-nelson** [comment]
> I have manually built and tested spl PR together with this change. All tests passed.

less concerned about the tests and more about what needs to happen to get the program redeployed with this patch

**Comment by t-nelson** [comment]
> > less concerned about the tests and more about what needs to happen to get the program redeployed with this patch
> 
> The current deserializers work properly with the change, so we don't actually _need_ to update the program. They just ignore the new stuff at the end of the struct

put yer money where yer mouth is and approve? :wink:

File: account-decoder/src/parse_stake.rs
Comment (2501babe) [inline account-decoder/src/parse_stake.rs:25]: im assuming the intent of flags (if more flags are ever added) is only for transient internal state tracking, so doesnt need to show up in `UiStakeAccount`

**Comment by 2501babe** [comment]
lgtm, also note i tested my single pool cli tests, with a cli built against this pr, and a program built against 1.16, and everything worked fine. i wasnt able to run any sbf tests for an issue unrelated to this pr (ie they fail on monorepo master too, i think because of the tokio upgrade)

i am curious what the upgrade story here is tho, will you cut a new monorepo release after this pr? or is this blocked on splitting sdk/program out of monorepo? since once this is merged, we would need something to update to, and in the mean time monorepo ci will break on the downstream projects tests

File: account-decoder/src/parse_stake.rs
Comment (t-nelson) [reply account-decoder/src/parse_stake.rs:25]: i don't think we need to impose any use-restrictions on the flags. future additions that make sense to display will likely need handling that we can't predict today. for instance i'll claim one when i get around proposing vote account bonds, but ui display would probably be something like
```rust
flags.contains(FLAG_VOTER_BOND).then_some(delegation.vote_account)
```

Comment (HaoranYi) [reply account-decoder/src/parse_stake.rs:25]: https://github.com/solana-labs/solana/pull/32033/files

^^ This is a follow up PR, which adds the stake flags to `UiStakeAccount`.

**Review by 2501babe — APPROVED** [review]

**Review by joncinque — COMMENTED** [review]
Late review, but just wanted to say that the stake program changes look good!

**Pull request merged**
